subprojects/docs/src/docs/userguide/lazyConfiguration.adoc

```diff
@@ -15,40 +15,226 @@
+* api:org.gradle.api.provider.Property[] are properties that can be queried and overwritten.
+** Properties with these types are configurable.
+** `Property` implements the `Provider` interface.
+** The method api:org.gradle.api.provider.Property#set(T)[] provides a value for the property, overwriting whatever value may have been present.
+** A `Property` can be created by the factory method api:org.gradle.api.model.ObjectFactory#property(java.lang.Class)[].
+
+Neither of these types nor their subtypes are intended to be implemented by a build script or plugin author.  Gradle provides several factory methods to create instances of these types. See the <<lazy_configuration_reference,Quick Reference>> for all of the types and factories available.
+
+Lazy properties are intended to be passed around and only evaluated when required (usually, during the execution phase). For more information about the Gradle build phases, please see <<sec:build_phases>>.
+
+The following demonstrates a task with a read-only property and a configurable property:
+
+++++
+<sample id="usePropertyAndProvider" dir="providers/propertyAndProvider" title="Using a read-only and configurable property">
+    <sourcefile file="build.gradle"/>
+    <output args="greeting"/>
+</sample>
+++++
+
+The `Greeting` task has a `Property<String>` for the mutable part of the message and a `Provider` for the modified, read-only, message.
+
+[NOTE]
+====
+
+Note that Groovy Gradle DSL will generate setter methods for each `Property`-typed property in a task implementation. These setter methods allow you to configure the property using the assignment (`=`) operator as a convenience.
+
+====
+
+[[sec:why_new_types]]
+=== Why new types?
+The Provider API is a public implementation of a feature Gradle has had internally for a long time--convention mapping. Provider types are faster, more understandable and better instrumented than the internal convention mapping mechanisms. The Provider API also enables more consistent build logic since it allows a plugin or task to express the relationship between properties without worrying about when particular values are determined.
+
+[[sec:creating_property_provider]]
+=== Creating a Property or Provider
+
+If provider types are not intended to be implemented directly by build script or plugin authors, how do you create new one? Gradle provides various factory APIs to create new instances of both api:org.gradle.api.provider.Provider[] and api:org.gradle.api.provider.Property[]:
+
+- api:org.gradle.api.provider.ProviderFactory#provider(java.util.concurrent.Callable)[] instantiates a new `Provider`. A handy method is available at api:org.gradle.api.Project#provider(java.util.concurrent.Callable)[]. If a api:org.gradle.api.Project[] instance isn’t available, you can inject api:org.gradle.api.provider.ProviderFactory[] through a constructor or method.
```

Comment:
If you provide different way to do the same thing then users often ask "which one is better" or "which one should I use". Maybe can give a concrete recommendation here.